### PR TITLE
Add VersionListDataClient to read and write the version list JSON

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/NuGetGallery/issues/6442.

Note that this is going into the `azuresearch` feature branch.

Example JSON:

```json
{
  "VersionProperties": {
    "1.0.0-beta.2": {
      "Listed": true,
      "SemVer2": true
    },
    "1.0.0-beta.10": {
      "Listed": true,
      "SemVer2": true
    },
    "2.0.0": {
      "Listed": true
    },
    "10.0.0": {}
  }
}
```

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/VersionList/VersionListData.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        public VersionListData(Dictionary<string, VersionPropertiesData> versionProperties)
+        {
+            VersionProperties = new Dictionary<string, VersionPropertiesData>(
+                versionProperties,
+                StringComparer.OrdinalIgnoreCase);
+        }
+
+        /// <summary>
+        /// A dictionary of all versions currently available for this package. This includes listed and unlisted
+        /// versions. The key is the full version string (can include build metadata). The look-up is case-insensitive,
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/VersionList/VersionListData.cs]: >The look-up is case-insensitive [](start = 87, length = 31)

Should we have a unit test on this?

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/VersionList/VersionListData.cs]: Oh, I meant to remove this behavior and only use `VersionListData` for serialization concerns.

I'll remove this.

File: src/NuGet.Services.AzureSearch/VersionList/VersionListDataClient.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        public async Task ReplaceAsync(string id, VersionListData data, IAccessCondition accessCondition)
+        {
+            using (var stream = new MemoryStream())
+            {
+                using (var streamWriter = new StreamWriter(
+                    stream,
+                    encoding: new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true),
+                    bufferSize: 1024,
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/VersionList/VersionListDataClient.cs]: Out of curiosity, what's the benefit of these encoding settings?

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/VersionList/VersionListDataClient.cs]: This is the default for `StreamWriter`, but I want to keep the stream open.

**Pull request merged**